### PR TITLE
Fix usage of deprecated MemorySize class in public API [HZ-1483]

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Fix usage of deprecated MemorySize class in public API.

`getSize`/`setSize` are now pointing to new property; both methods are deprecated. XML and YAML parsing should still read `size` property without an issue.

EE part: [5315](https://github.com/hazelcast/hazelcast-enterprise/pull/5315)

Fixes HZ-1483

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by blazember** [comment]
Thanks @TomaszGaweda :+1: Have you checked the declarative config as well to make sure we don't have a similar gotcha there? It seems in the xsd we define the inheritance the other way around, so capacity is defined with type `memory-size`. It won't be noticed by a user most likely though, just hurting my love for symmetry :smile:

**Comment by TomaszGaweda** [comment]
@blazember I didn't change XSDs to not break compatibility; I cannot distinguish getter from setter and/or use generics

**Review by blazember — APPROVED** [review]

**Comment by frant-hartm** [comment]
run-lab-run

**Comment by TomaszGaweda** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
Unfortunately, this won't do the trick for binary compatibility. Assuming this existing client code:

```
MemorySize memSize = config.getNativeMemoryConfig().getSize();

javap:
...
INVOKEVIRTUAL com/hazelcast/config/NativeMemoryConfig.getSize ()Lcom/hazelcast/memory/MemorySize;
...
```

But the new signature of the method is actually:
```
public <T extends com.hazelcast.memory.Capacity> T getSize();
    descriptor: ()Lcom/hazelcast/memory/Capacity;
    flags: ACC_PUBLIC
```

so above code compiled against 5.1 and executed with 5.2.0-SNAPSHOT in classpath will throw
```
Exception in thread "main" java.lang.NoSuchMethodError: 'com.hazelcast.memory.MemorySize com.hazelcast.config.NativeMemoryConfig.getSize()'
```

 It would work with bridge methods but javac only produces those under specific circumstances. According to https://www.morling.dev/blog/refining-return-type-java-methods-without-breaking-backwards-compatibility/ there are some tools to introduce bridge methods (require some form of bytecode manipulation). wdyt?

**Comment by TomaszGaweda** [comment]
Seems to be working @vbekiaris :)

File: hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java [hunk 1]
```diff
@@ -67,7 +68,7 @@ public class NativeMemoryConfig {
     public static final int INITIAL_MEMORY_SIZE = MIN_INITIAL_MEMORY_SIZE;
 
     private boolean enabled;
-    private MemorySize size = new MemorySize(INITIAL_MEMORY_SIZE, MemoryUnit.MEGABYTES);
+    private Capacity size = new Capacity(INITIAL_MEMORY_SIZE, MemoryUnit.MEGABYTES);
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java:71]: I don't have a strong opinion about this, but wouldn't the following lead to a cleaner API without hacks?

- rename the field to capacity, introduce new getter/setter for that
- introduce the new field to the xml/yaml/spring configs (this is the annoying bit, but it's easier because the NativeMemoryConfig is not dynamic, so no ser/de in client protocol)
- deprecate getSize/setSize methods, keep their signature (hence binary compatibility), and internally change the capacity field
- `memory-size` in xsd can be renamed to capacity, leaving one deprecated config with `<xs:element name="size" type="capacity" minOccurs="0"/>`, reversing the symmetry (now we have 3 `capacity` fields with type `memory-size`)

wdyt?

Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java:71]: @TomaszGaweda I am fine with both approaches (bridge method or add new `get/setCapacity`).

Notice that if you opt for the former, we need to keep `NativeMemoryConfig setSize(MemorySize)` method to maintain binary compat.

File: hazelcast/src/main/resources/hazelcast-config-5.2.xsd [hunk 2]
```diff
@@ -3313,7 +3313,8 @@
 
     <xs:complexType name="native-memory">
         <xs:all>
-            <xs:element name="size" type="memory-size" minOccurs="0"/>
+            <xs:element name="size" type="capacity" minOccurs="0"/>
```
Comment (vbekiaris) [inline hazelcast/src/main/resources/hazelcast-config-5.2.xsd:3316]: Maybe add an `xs:annotation/xs:documentation` child element stating this is deprecated and indicating `capacity` as the element to use in the future? (similarly elsewhere)

**Review by vbekiaris — APPROVED** [review]
Looks good, added a suggestion

File: hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java [hunk 3]
```diff
@@ -103,11 +106,35 @@ public MemorySize getSize() {
...
      * with {@link NativeOutOfMemoryError}.
      *
-     * @param size memory size
+     * @param capacity memory size
+     * @return this {@link NativeMemoryConfig} instance
+     * @deprecated Since 5.2, use {@link #setCapacity(Capacity)} instead.
+     */
+    @Deprecated
+    public NativeMemoryConfig setSize(MemorySize capacity) {
+        this.capacity = isNotNull(capacity, "size");
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java:115]: If have 2 instances, one sets the `capacity` via `setCapacity`, the other via `setSize`, are they still equal??

Maybe add a test into `NativeMemoryConfigTest`

File: hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java [hunk 4]
```diff
@@ -103,11 +106,35 @@ public MemorySize getSize() {
...
+     * Returns size (capacity) of the native memory region.
+     */
+    public Capacity getCapacity() {
+        return capacity;
+    }
+
+    /**
+     * Sets size (capacity) of the native memory region.
+     * <p>
+     * Total size of the memory blocks allocated in native memory region cannot exceed this memory size.
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java:129]: `size` -> `capacity`?

File: hazelcast/src/main/resources/hazelcast-config-5.1.json [hunk 5]
```diff
@@ -1497,7 +1497,7 @@
                 "additionalProperties": false,
                 "properties": {
                   "capacity": {
-                    "$ref": "#/definitions/MemorySize",
+                    "$ref": "#/definitions/Capacity",
```
Comment (frant-hartm) [inline hazelcast/src/main/resources/hazelcast-config-5.1.json:1500]: This is `hazelcast-config-5.1.json`, is this change a mistake?

File: hazelcast/src/main/resources/hazelcast-config-5.2.xsd [hunk 6]
```diff
@@ -3313,7 +3313,8 @@
...
+            <xs:element name="capacity" type="capacity" minOccurs="0"/>
```
Comment (frant-hartm) [inline hazelcast/src/main/resources/hazelcast-config-5.2.xsd:3317]: What if there are both parameters?

File: hazelcast/src/test/java/com/hazelcast/config/NativeMemoryConfigTest.java [hunk 7]
```diff
@@ -1,41 +0,0 @@
...
-public class NativeMemoryConfigTest {
-
-    @Test
-    public void testEqualsAndHashCode() {
-        assumeDifferentHashCodes();
-        EqualsVerifier.forClass(NativeMemoryConfig.class)
-                      .suppress(Warning.NULL_FIELDS, Warning.NONFINAL_FIELDS)
-                      .verify();
-    }
-}
```
Comment (frant-hartm) [inline hazelcast/src/test/java/com/hazelcast/config/NativeMemoryConfigTest.java:41]: Why did you delete this?

File: hazelcast/src/test/java/com/hazelcast/internal/memory/MemorySizeTest.java [hunk 8]
```diff
@@ -1,42 +0,0 @@
...
-public class MemorySizeTest {
-
-    @Test
-    public void testEqualsAndHashCode() {
-        assumeDifferentHashCodes();
-        EqualsVerifier.forClass(MemorySize.class)
-                .suppress(Warning.NONFINAL_FIELDS)
-                .verify();
-    }
-}
```
Comment (frant-hartm) [inline hazelcast/src/test/java/com/hazelcast/internal/memory/MemorySizeTest.java:42]: MemorySize is deprecated, but we shouldn't delete the test.

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/resources/hazelcast-config-5.2.xsd (hunk 6 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/resources/hazelcast-config-5.2.xsd:3317]: Last one will be picked. TBH, if you have any other idea how to keep XSD backward compatible and up-to-date with Java config, I'd be super grateful, I don't like this repetition as well :P

File: hazelcast/src/test/java/com/hazelcast/internal/memory/MemorySizeTest.java (hunk 8 above)
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/internal/memory/MemorySizeTest.java]: 316de23b436bfd2578cf7ef48ea185be80dcfe23

File: hazelcast/src/test/java/com/hazelcast/config/NativeMemoryConfigTest.java (hunk 7 above)
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/config/NativeMemoryConfigTest.java]: I don't really remember, probably by mistake... 316de23b436bfd2578cf7ef48ea185be80dcfe23

File: hazelcast/src/main/resources/hazelcast-config-5.1.json (hunk 5 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/resources/hazelcast-config-5.1.json]: Yeah, changed should be in 5.2 only. 316de23b436bfd2578cf7ef48ea185be80dcfe23

File: hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java (hunk 4 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java]: 316de23b436bfd2578cf7ef48ea185be80dcfe23

File: hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java (hunk 3 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/java/com/hazelcast/config/NativeMemoryConfig.java:115]: Yes, they are equal. Added test in 316de23b436bfd2578cf7ef48ea185be80dcfe23

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java [hunk 9]
```diff
@@ -1387,7 +1387,7 @@ private void handleMemoryTierConfig(BeanDefinitionBuilder configBuilder, Node no
         }
 
         private void handleCapacity(Node node, BeanDefinitionBuilder nativeMemoryConfigBuilder) {
-            BeanDefinitionBuilder memorySizeConfigBuilder = createBeanBuilder(MemorySize.class);
+            BeanDefinitionBuilder memorySizeConfigBuilder = createBeanBuilder(Capacity.class);
```
Comment (mdumandag) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java:1390]: minor: maybe change the variable name to capacity as well

File: hazelcast/src/main/resources/hazelcast-config-5.2.xsd (hunk 2 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/resources/hazelcast-config-5.2.xsd]: Good idea, 074f8ffed1dbc57d44e8a20984538bb6591851cb

File: hazelcast/src/main/resources/hazelcast-config-5.1.json [hunk 10]
```diff
@@ -2940,7 +2940,7 @@
         "migration"
       ]
     },
-    "MemorySize": {
+    "Capacity": {
```
Comment (mdumandag) [inline hazelcast/src/main/resources/hazelcast-config-5.1.json:2943]: this is changing the 5.1 json schema

File: hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java [hunk 11]
```diff
@@ -665,6 +668,20 @@ public void testPersistentMemoryConfiguration_NotExistingModeThrows() {
         buildConfig(xml);
     }
 
+    @Test
+    @Override
+    public void testNativeMemoryConfiguration_isBackwardCompatible() {
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java:673]: Should we also change the `nativeMemory` test in this file to use the new capacity field, instead of size?

File: hazelcast/src/test/java/com/hazelcast/client/config/YamlClientConfigBuilderTest.java [hunk 12]
```diff
@@ -665,6 +668,21 @@ public void testPersistentMemoryConfiguration_NotExistingModeThrows() {
         buildConfig(yaml);
     }
 
+    @Test
+    @Override
+    public void testNativeMemoryConfiguration_isBackwardCompatible() {
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/client/config/YamlClientConfigBuilderTest.java:673]: similar comment to yaml test file above

File: hazelcast/src/test/java/com/hazelcast/config/XMLConfigBuilderTest.java [hunk 13]
```diff
@@ -3395,7 +3394,7 @@ public void testLocalDevice() {
         assertEquals(blockSize * device0Multiplier, localDeviceConfig.getBlockSize());
         assertEquals(readIOThreadCount * device0Multiplier, localDeviceConfig.getReadIOThreadCount());
         assertEquals(writeIOThreadCount * device0Multiplier, localDeviceConfig.getWriteIOThreadCount());
-        assertEquals(new MemorySize(1234567890, MemoryUnit.MEGABYTES), localDeviceConfig.getCapacity());
+        assertEquals(new Capacity(1234567890, MemoryUnit.MEGABYTES), localDeviceConfig.getCapacity());
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/config/XMLConfigBuilderTest.java:3397]: Should we have a test similar to `testNativeMemoryConfiguration_isBackwardCompatible` here as well?

File: hazelcast/src/test/java/com/hazelcast/config/YamlConfigBuilderTest.java [hunk 14]
```diff
@@ -3360,7 +3359,7 @@ public void testLocalDevice() {
         assertEquals(blockSize * device0Multiplier, localDeviceConfig.getBlockSize());
         assertEquals(readIOThreadCount * device0Multiplier, localDeviceConfig.getReadIOThreadCount());
         assertEquals(writeIOThreadCount * device0Multiplier, localDeviceConfig.getWriteIOThreadCount());
-        assertEquals(new MemorySize(1234567890, MemoryUnit.MEGABYTES), localDeviceConfig.getCapacity());
+        assertEquals(new Capacity(1234567890, MemoryUnit.MEGABYTES), localDeviceConfig.getCapacity());
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/config/YamlConfigBuilderTest.java:3362]: Should we have a test similar to `testNativeMemoryConfiguration_isBackwardCompatible` here as well?

File: hazelcast/src/main/resources/hazelcast-config-5.2.xsd (hunk 6 above)
Comment (frant-hartm) [reply hazelcast/src/main/resources/hazelcast-config-5.2.xsd]: I am fine with last one wins.

File: hazelcast/src/test/java/com/hazelcast/config/YamlConfigBuilderTest.java (hunk 14 above)
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/config/YamlConfigBuilderTest.java:3362]: It was just removal of usages of deprecated class, here field is the same (already `capacity`, no legacy `size` field)

File: hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java (hunk 11 above)
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java:673]: 1135d7cd77ef1592681bdc17a0779a1f4d21915e

File: hazelcast/src/test/java/com/hazelcast/client/config/YamlClientConfigBuilderTest.java (hunk 12 above)
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/client/config/YamlClientConfigBuilderTest.java:673]: 1135d7cd77ef1592681bdc17a0779a1f4d21915e

File: hazelcast/src/main/resources/hazelcast-config-5.1.json (hunk 10 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/resources/hazelcast-config-5.1.json]: Yep,  by mistake. Reverted: 1135d7cd77ef1592681bdc17a0779a1f4d21915e

**Comment by TomaszGaweda** [comment]
run-lab-run

**Comment by TomaszGaweda** [comment]
run-lab-run

**Review by frant-hartm — APPROVED** [review]
